src/java.base/share/classes/sun/security/provider/ML_DSA.java

```diff
@@ -0,0 +1,1427 @@
+                if (rplus - r0 == mlDsa_q - 1) {
+                    r1 = 0;
+                    r0 = r0 - 1;
+                } else {
+                    r1 = (rplus - r0) / (2*gamma2);
+                }
+                lowPart[i][m] = r0;
+                highPart[i][m] = r1;
+            }
+        }
+    }
+
+    private int highBits(int input) {
+        int rplus = (input + mlDsa_q) % mlDsa_q;
+        int r0 = rplus % (2*gamma2);
+        r0 -= r0 > gamma2 ? 2*gamma2 : 0;
+        int r1;
+        if (rplus - r0 == mlDsa_q - 1) {
+            r1 = 0;
+            r0 = r0 - 1;
+        } else {
+            r1 = (rplus - r0) / (2*gamma2);
+        }
+        return r1;
+    }
+
+    private boolean[][] makeHint(int[][] z, int[][] r) {
+        int[][] v1 = vectorAdd(r,z);
+        boolean[][] res = new boolean[mlDsa_k][mlDsa_n];
+        for (int i = 0; i < mlDsa_k; i++) {
+            for (int j = 0; j < mlDsa_n; j++) {
+                int r1High = highBits(r[i][j]);
+                int v1High = highBits(v1[i][j]);
+                res[i][j] = r1High != v1High;
+            }
+        }
+        return res;
+    }
+
+    private int[][] useHint(boolean[][] h, int[][] r) {
+        int m = (mlDsa_q - 1) / (2*gamma2);
+        int[][] lowPart = new int[mlDsa_k][mlDsa_n];
+        int[][] highPart = new int[mlDsa_k][mlDsa_n];
+        decompose(r, lowPart, highPart);
+
+        for (int i = 0; i < mlDsa_k; i++) {
+            for (int j = 0; j < mlDsa_n; j++) {
+                if (h[i][j]) {
+                    highPart[i][j] += lowPart[i][j] > 0 ? 1 : -1;
+                }
+                highPart[i][j] = ((highPart[i][j] % m) + m) % m;
+            }
+        }
+        return highPart;
+    }
+
+    /*
+    NTT functions as specified in Section 7.5 of specification
+    */
+
+    public static int[] mlDsaNtt(int[] coeffs) {
+        int result = implMlDsaAlmostNtt(coeffs, montZetasForVectorNtt);
+        int[] check = coeffs.clone();
+        result = implMlDsaMontMulByConstant(coeffs,  montRModQ);
+        return coeffs;
+    }
+
+    @IntrinsicCandidate
+    static int implMlDsaAlmostNtt(int[] coeffs, int[] zetas) {
+        return implMlDsaAlmostNttJava(coeffs);
+    }
+
+    static int implMlDsaAlmostNttJava(int[] coeffs) {
+        int dimension = mlDsa_n;
+        int m = 0;
+        for (int l = dimension / 2; l > 0; l /= 2) {
+            for (int s = 0; s < dimension; s += 2 * l) {
+                for (int j = s; j < s + l; j++) {
+                    int tmp = montMul(montZetasForNtt[m], coeffs[j + l]);
+                    coeffs[j + l] = coeffs[j] - tmp;
+                    coeffs[j] = coeffs[j] + tmp;
+                }
+                m++;
+            }
+        }
+
+        return 1;
+    }
+
+    public static int[] mlDsaInverseNtt(int[] coeffs) {
+        int result = implMlDsaAlmostInverseNtt(coeffs, montZetasForVectorInverseNtt);
+        result = implMlDsaMontMulByConstant(coeffs, montDimInverse);
```

Comment:
In FIPS 204, the constant is 8347681. Why do you use 16382?

Reply:
The same thing is happening here as with the NTT, except we are not multiplying with 1 but 2^-256 using Montgomery multiplication combined with the conversion back to "normal domain", that is why the variable name is montDimInverse. In the standard, they have dinInverse.